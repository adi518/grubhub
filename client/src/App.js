import React from 'react';
import { useEffectOnce } from 'react-use';

import { useStore } from './store';
import { parseQueryString } from './utils';
import { getToken, getTweets } from './api';

import StyledApp from './App.styled';
import Logo from './components/Logo';
import Tweet from './components/Tweet';
import Tweets from './components/Tweets';
import TwitterLogo from './components/TwitterLogo';
import Loader, { SmallLoader } from './components/Loader';

function App({ initialState }) {
  const { state, setToken, setIsLoadingMore, setTweets } = useStore(
    initialState
  );

  useEffectOnce(() => {
    (async () => {
      const {
        data: { access_token: accessToken }
      } = await getToken();
      setToken(accessToken);
      const {
        data: {
          statuses: tweets,
          search_metadata: { next_results: nextResults }
        }
      } = await getTweets({ accessToken });
      setTweets({ tweets, nextResults });
    })();
  });

  const handleScroll = async e => {
    const clientHeight = parseInt(e.target.clientHeight * 1.667);
    const scrollEnd =
      e.target.scrollHeight - e.target.scrollTop <= clientHeight;
    if (scrollEnd && state.nextResults && !state.isLoadingMore) {
      setIsLoadingMore(true);
      const {
        data: {
          statuses: tweets,
          search_metadata: { next_results: nextResults }
        }
      } = await getTweets({
        accessToken: state.token,
        query: parseQueryString(state.nextResults)
      });
      setIsLoadingMore(false);
      setTweets({ tweets, nextResults });
    }
  };

  return (
    <StyledApp onScroll={handleScroll}>
      <Logo />
      <TwitterLogo />
      {state.isLoadingMore && <SmallLoader />}
      {state.tweets.length ? (
        <Tweets>
          <React.Fragment>
            {state.tweets.map(tweet => (
              <Tweet {...tweet} key={tweet.id} />
            ))}
          </React.Fragment>
        </Tweets>
      ) : (
        <Loader />
      )}
    </StyledApp>
  );
}

export default App;
