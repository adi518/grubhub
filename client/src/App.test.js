import React from 'react';
import { render } from '@testing-library/react';
import ShallowRenderer from 'react-test-renderer/shallow';

import App from './App';
import tweetsMock from './tweetsMock';

const renderer = new ShallowRenderer();
const {
  statuses: [mockTweet0, mockTweet1, mockTweet2]
} = tweetsMock;

describe('App component', () => {
  test('renders without tweets', () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders with tweets', () => {
    const tweets = [mockTweet0, mockTweet1, mockTweet2];
    const { asFragment } = render(<App initialState={{ tweets }} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
