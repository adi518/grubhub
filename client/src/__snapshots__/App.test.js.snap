// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders with tweets 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa pKMFB"
  >
    <div
      class="sc-gzVnrw cjsdOt"
    >
      <a
        href="https://www.github.com/"
      >
        <img
          alt="Octocat"
          src="foundingfather.png"
        />
      </a>
    </div>
    <div
      class="sc-htoDjs glFZZT"
    >
      <a
        href="https://twitter.com/"
      >
        <svg
          viewBox="0 0 24 24"
        >
          <g>
            <path
              d="M23.643 4.937c-.835.37-1.732.62-2.675.733.962-.576 1.7-1.49 2.048-2.578-.9.534-1.897.922-2.958 1.13-.85-.904-2.06-1.47-3.4-1.47-2.572 0-4.658 2.086-4.658 4.66 0 .364.042.718.12 1.06-3.873-.195-7.304-2.05-9.602-4.868-.4.69-.63 1.49-.63 2.342 0 1.616.823 3.043 2.072 3.878-.764-.025-1.482-.234-2.11-.583v.06c0 2.257 1.605 4.14 3.737 4.568-.392.106-.803.162-1.227.162-.3 0-.593-.028-.877-.082.593 1.85 2.313 3.198 4.352 3.234-1.595 1.25-3.604 1.995-5.786 1.995-.376 0-.747-.022-1.112-.065 2.062 1.323 4.51 2.093 7.14 2.093 8.57 0 13.255-7.098 13.255-13.254 0-.2-.005-.402-.014-.602.91-.658 1.7-1.477 2.323-2.41z"
            />
          </g>
        </svg>
      </a>
    </div>
    <div
      class="sc-bZQynM iJNgEP"
    >
      <div
        class="sc-bxivhb jtrJuU"
      >
        <div
          class="sc-ifAKCX WOpjk"
        >
          <div
            class="sc-htpNat gwmiLf"
          >
            <img
              alt="josephus1"
              src="http://abs.twimg.com/sticky/default_profile_images/default_profile_normal.png"
            />
          </div>
          <div>
            <b>
              @josephus1
            </b>
            <div
              class="sc-EHOje xTGQR"
            >
              RT @chrismartenson: Today's #coronavirus #COVID2019  VIDEO update is out.
I explore and discuss the days events and the explosive conclusio…
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-bxivhb jtrJuU"
      >
        <div
          class="sc-ifAKCX WOpjk"
        >
          <div
            class="sc-htpNat gwmiLf"
          >
            <img
              alt="Checkmate"
              src="http://pbs.twimg.com/profile_images/1222673174329880576/6LLEIcbW_normal.jpg"
            />
          </div>
          <div>
            <b>
              @Checkmate
            </b>
            <div
              class="sc-EHOje xTGQR"
            >
              RT @BeThePlan: What if [they] are trying to kill off as many people as possible, anyhow [they] can, while [they] still can... what if it’s…
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-bxivhb jtrJuU"
      >
        <div
          class="sc-ifAKCX WOpjk"
        >
          <div
            class="sc-htpNat gwmiLf"
          >
            <img
              alt="Rich Dunsheath"
              src="http://pbs.twimg.com/profile_images/386429619/Mean_Daddy_sm_normal.jpg"
            />
          </div>
          <div>
            <b>
              @Rich Dunsheath
            </b>
            <div
              class="sc-EHOje xTGQR"
            >
              RT @SecPompeo: The U.S. is deeply concerned about the vulnerability of the North Korean people to a #coronavirus outbreak. We strongly supp…
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App component renders without tweets 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa pKMFB"
  >
    <div
      class="sc-gzVnrw cjsdOt"
    >
      <a
        href="https://www.github.com/"
      >
        <img
          alt="Octocat"
          src="foundingfather.png"
        />
      </a>
    </div>
    <div
      class="sc-htoDjs glFZZT"
    >
      <a
        href="https://twitter.com/"
      >
        <svg
          viewBox="0 0 24 24"
        >
          <g>
            <path
              d="M23.643 4.937c-.835.37-1.732.62-2.675.733.962-.576 1.7-1.49 2.048-2.578-.9.534-1.897.922-2.958 1.13-.85-.904-2.06-1.47-3.4-1.47-2.572 0-4.658 2.086-4.658 4.66 0 .364.042.718.12 1.06-3.873-.195-7.304-2.05-9.602-4.868-.4.69-.63 1.49-.63 2.342 0 1.616.823 3.043 2.072 3.878-.764-.025-1.482-.234-2.11-.583v.06c0 2.257 1.605 4.14 3.737 4.568-.392.106-.803.162-1.227.162-.3 0-.593-.028-.877-.082.593 1.85 2.313 3.198 4.352 3.234-1.595 1.25-3.604 1.995-5.786 1.995-.376 0-.747-.022-1.112-.065 2.062 1.323 4.51 2.093 7.14 2.093 8.57 0 13.255-7.098 13.255-13.254 0-.2-.005-.402-.014-.602.91-.658 1.7-1.477 2.323-2.41z"
            />
          </g>
        </svg>
      </a>
    </div>
    <div
      class="sc-dnqmqq dpfhSX"
    >
      <div
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate"
        role="progressbar"
        style="width: 100px; height: 100px;"
      >
        <svg
          class="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
